src/AudioCaptureImpl_WASAPI.cpp

```diff
@@ -1,25 +1,631 @@
+}
+
+ULONG AudioCaptureImpl::Release()
+{
+    return InterlockedDecrement(&_referenceCount);
+}
+
+std::string AudioCaptureImpl::UnicodeToString(LPCWSTR unicodeString)
+{
+    std::string utf8String;
+    Poco::UnicodeConverter::convert(std::wstring(unicodeString), utf8String);
+    return utf8String;
+}
+
+std::vector<AudioCaptureImpl::AudioDevice> AudioCaptureImpl::GetAudioDeviceList(IMMDeviceEnumerator* enumerator) const
+{
+    auto addEndpoints = [this, enumerator](std::vector<AudioDevice>& deviceList, EDataFlow dataFlow) {
+        HRESULT result{S_OK};
+        IMMDeviceCollection* audioEndpoints{nullptr};
+
+        result = enumerator->EnumAudioEndpoints(dataFlow, DEVICE_STATE_ACTIVE, &audioEndpoints);
+        if (FAILED(result))
+        {
+            poco_error_f1(_logger, "IMMDeviceEnumerator::EnumAudioEndpoints failed: result = 0x%08?x", result);
+            return;
+        }
+
+        UINT deviceCount{0};
+        audioEndpoints->GetCount(&deviceCount);
+        for (UINT item = 0; item < deviceCount; item++)
+        {
+            IMMDevice* device{nullptr};
+            result = audioEndpoints->Item(item, &device);
+
+            if (FAILED(result) || device == nullptr)
+            {
+                poco_error_f2(_logger, "IMMDeviceEnumerator::Item failed for device %?u: result = 0x%08?x", item, result);
+                continue;
+            }
+
+            deviceList.emplace_back(device, (dataFlow == eRender));
+            device->Release();
+        }
+
+        audioEndpoints->Release();
+    };
+
+    std::vector<AudioDevice> deviceList;
+    addEndpoints(deviceList, eRender);
+    addEndpoints(deviceList, eCapture);
+
+    return deviceList;
+}
+
+bool AudioCaptureImpl::OpenAudioDevice(IMMDevice* device, bool useLoopback)
+{
+    // activate an IAudioClient
+    HRESULT result = device->Activate(__uuidof(IAudioClient),
+                              CLSCTX_ALL,
+                              nullptr,
+                              reinterpret_cast<void**>(&_audioClient));
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IMMDevice::Activate(IAudioClient) failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    // get the default device periodicity
+    REFERENCE_TIME hnsDefaultDevicePeriod;
+    result = _audioClient->GetDevicePeriod(&hnsDefaultDevicePeriod,
+                                           nullptr);
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IAudioClient::GetDevicePeriod failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    // get the default device format
+    WAVEFORMATEX* pwfx;
+    result = _audioClient->GetMixFormat(&pwfx);
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IAudioClient::GetMixFormat failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    // Should default to float32 data, but some devices might deliver other formats.
+    if (pwfx->wFormatTag != WAVE_FORMAT_IEEE_FLOAT)
```

Comment:
There is a way to have WASAPI convert the sample format to float if it's something else, e.g. 16 bit int/uint samples. I've tried many devices, and could not find any that resulted in a different sample format, but some special/ancient recording devices could still deliver a different format.

If we get frequent user reports about this error, there's some room for improvement. The original SDL app did that, but converting float samples to UINT16 (and then right back to float inside libprojectM).